src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,114 @@
+        /// that this is bool and not bool? because we know that the symbol is at the very least declared, 
+        /// so there's no need to return a null value. 
+        /// </summary>
+        public static bool IsDeclaredSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, SyntaxNode declarationSyntax, CancellationToken cancellationToken)
+        {
+            var declaredType = semanticModel.GetRequiredDeclaredSymbol(declarationSyntax, cancellationToken);
+            var declaredOperation = semanticModel.GetRequiredOperation(declarationSyntax, cancellationToken);
+
+            var rootOperation = declaredOperation;
+
+            // Walk up the tree to find a root for the operation
+            // that contains the declaration
+            while (rootOperation.Parent is not null)
+            {
+                if (rootOperation is IBlockOperation)
+                {
+                    break;
+                }
+
+                rootOperation = rootOperation.Parent;
+            }
+
+            return IsSymbolAssignedPossiblyNullValue(semanticModel, rootOperation, declaredType) == true;
+        }
+
+        /// <summary>
+        /// Given an operation, goes through all decendent operations and returns true if the symbol passed in
+        /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+        /// null if no references are found, letting the caller determine what to do with that information
+        /// </summary>
+        public static bool? IsSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, IOperation operation, ISymbol symbol)
+        {
+            var references = operation.DescendantsAndSelf()
+                .Where(o => IsSymbolReferencedByOperation(o, symbol, allowNullInitializer: false));
+
+            var hasReference = false;
+
+            foreach (var reference in references)
+            {
+                hasReference = true;
+
+                // foreach statements are handled special because the iterator is not assignable, so the elementtype 
+                // annotation is accurate for determining if the loop declaration has a reference that allows the symbol
+                // to be null
+                if (reference is IForEachLoopOperation forEachLoop)
+                {
+                    var foreachInfo = semanticModel.GetForEachStatementInfo((CommonForEachStatementSyntax)forEachLoop.Syntax);
+
+                    if (foreachInfo.ElementType is null)
+                    {
+                        continue;
+                    }
+
+                    // Use NotAnnotated here to keep both Annotated and None (oblivious) treated the same, since
+                    // this is directly looking at the annotation and not the flow state
+                    if (foreachInfo.ElementType.NullableAnnotation != NullableAnnotation.NotAnnotated)
+                    {
+                        return true;
```

Comment:
This isn't checking if there's another reference somewhere else?

Reply:
You can't reassign in a foreach loop so the original annotation should be correct

Reply:
Ah right, hmm, I think I was trying to figure out if something like:

`foreach (string? s in GetNonNullableStrings()) { }`

should returning true or "null" in this case, since nothing is actually referencing this?